### PR TITLE
pallet-treasury | port to frame v2

### PR DESCRIPTION
**Pull request by shamb0**
# pallet-treasury | port to frame v2

related #7882

polkadot companion: https://github.com/paritytech/polkadot/pull/2730

:warning: Breaking Change :warning:

From https://crates.parity.io/frame_support/attr.pallet.html#checking-upgrade-guidelines

```
storages now use PalletInfo for module_prefix instead of the one given to decl_storage: 
Thus any use of this pallet in construct_runtime! should be careful to update name 
in order not to break storage or to upgrade storage (moreover for instantiable pallet). 
If pallet is published, make sure to warn about this breaking change.
```

So users of the treasury pallet must be careful about the name they used in construct_runtime!. Hence the runtime-migration label, which might not be needed depending on the configuration of the treasury pallet.

### CODE REVIEW

**Comment by thiolliere** [comment]
some tests in bounties needs to be fix it seems, because treasury no longer export DefaultInstance.
Instead bounties should use the default instance `()` or just removing the instance generic

**Comment by thiolliere** [comment]
would be good to make use of `frame_support::storage::migration::move_storage_from_pallet` in a migration function to move all storages, and then call this function either here in on_runtime_upgrade or in a polkadot companion at the runtime level

**Comment by shamb0** [comment]
> would be good to make use of `frame_support::storage::migration::move_storage_from_pallet` in a migration function to move all storages, and then call this function either here in on_runtime_upgrade or in a polkadot companion at the runtime level

Thanks @thiolliere , will pick up for this week

File: frame/treasury/src/lib.rs [hunk 1]
```diff
@@ -168,108 +135,85 @@ pub struct Proposal<AccountId, Balance> {
...
 
-decl_module! {
-	pub struct Module<T: Config<I>, I: Instance=DefaultInstance>
-		for enum Call
-		where origin: T::Origin
-	{
 		/// Fraction of a proposal's value that should be bonded in order to place the proposal.
 		/// An accepted proposal gets these back. A rejected proposal does not.
-		const ProposalBond: Permill = T::ProposalBond::get();
+		type ProposalBond: Get<Permill>;
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:167]:
```suggestion
		#[pallet::constant]
		type ProposalBond: Get<Permill>;
```

File: frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -168,108 +135,85 @@ pub struct Proposal<AccountId, Balance> {
...
 
 		/// Minimum amount of funds that should be placed in a deposit for making a proposal.
-		const ProposalBondMinimum: BalanceOf<T, I> = T::ProposalBondMinimum::get();
+		type ProposalBondMinimum: Get<BalanceOf<Self, I>>;
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:170]:
```suggestion
		#[pallet::constant]
		type ProposalBondMinimum: Get<BalanceOf<Self, I>>;
```

File: frame/treasury/src/lib.rs [hunk 3]
```diff
@@ -168,108 +135,85 @@ pub struct Proposal<AccountId, Balance> {
...
 
 		/// Period between successive spends.
-		const SpendPeriod: T::BlockNumber = T::SpendPeriod::get();
+		type SpendPeriod: Get<Self::BlockNumber>;
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:173]:
```suggestion
		#[pallet::constant]
		type SpendPeriod: Get<Self::BlockNumber>;
```

File: frame/treasury/src/lib.rs [hunk 4]
```diff
@@ -168,108 +135,85 @@ pub struct Proposal<AccountId, Balance> {
...
 
 		/// Percentage of spare funds (if any) that are burnt per spend period.
-		const Burn: Permill = T::Burn::get();
+		type Burn: Get<Permill>;
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:176]:
```suggestion
		#[pallet::constant]
		type Burn: Get<Permill>;
```

File: frame/treasury/src/lib.rs [hunk 5]
```diff
@@ -168,108 +135,85 @@ pub struct Proposal<AccountId, Balance> {
...
-				let _ = T::Currency::make_free_balance_be(
-					&account_id,
-					min,
-				);
-			}
-		});
-	}
-}
+		/// The treasury's module id, used for deriving its sovereign account ID.
+		type ModuleId: Get<ModuleId>;
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:148]:
```suggestion
		#[pallet::constant]
		type ModuleId: Get<ModuleId>;
```

File: frame/treasury/src/lib.rs [hunk 6]
```diff
@@ -278,23 +222,32 @@ decl_module! {
...
 
 			let c = Self::proposal_count();
-			<ProposalCount<I>>::put(c + 1);
-			<Proposals<T, I>>::insert(c, Proposal { proposer, value, beneficiary, bond });
-
-			Self::deposit_event(RawEvent::Proposed(c));
+			<ProposalCount<T, I>>::put(c + 1);
+			<Proposals<T, I>>::insert(
+				c,
+				Proposal::<T::AccountId, BalanceOf<T, I>>{
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:228]:
```suggestion
				Proposal::<T::AccountId, BalanceOf<T, I>> {
```

File: frame/treasury/src/lib.rs [hunk 7]
```diff
@@ -278,23 +222,32 @@ decl_module! {
...
-		pub fn reject_proposal(origin, #[compact] proposal_id: ProposalIndex) {
+		#[pallet::weight(
+			(
+				T::WeightInfo::reject_proposal(),
+				DispatchClass::Operational,
+			)
+		)]
+		pub fn reject_proposal(
+			origin: OriginFor<T>,
+			#[pallet::compact] proposal_id: ProposalIndex
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:249]:
```suggestion
			#[pallet::compact] proposal_id: ProposalIndex,
```

File: frame/treasury/src/lib.rs [hunk 8]
```diff
@@ -303,46 +256,137 @@ decl_module! {
...
+		Burnt(BalanceOf<T, I>),
+		/// Spending has finished; this is the amount that rolls over until next spend.
+		/// \[budget_remaining\]
+		Rollover(BalanceOf<T, I>),
+		/// Some funds have been deposited. \[deposit\]
+		Deposit(BalanceOf<T, I>),
+	}
+
+	/// Old name generated by `decl_event`.
+	// TODO :: Have to re-check is "Note message" required ?
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:305]:
```suggestion
```

**Review by thiolliere — COMMENTED** [review]
you need to compare the metadata to catch the miss like the constant.
Also master needs to be merged.

**Comment by shamb0** [comment]
Thanks @thiolliere, pulled in review suggested comments & merged to master :+1:

**Comment by thiolliere** [comment]
CI is failing https://gitlab.parity.io/parity/substrate/-/jobs/868957

**Comment by shamb0** [comment]
Hello @thiolliere, extremely, sorry. Overlooked at CI builds & dependency. today will have a look on "continuous-integration/gitlab-check-polkadot-companion-build"

File: frame/tips/src/tests.rs [hunk 9]
```diff
@@ -42,7 +42,7 @@ frame_support::construct_runtime!(
 	{
 		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
 		Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
-		Treasury: pallet_treasury::{Pallet, Call, Storage, Config, Event<T>},
+        Treasury: pallet_treasury::{Pallet, Call, Storage, Config<T>, Event<T>},
```
Comment (thiolliere) [inline frame/tips/src/tests.rs:45]: weird indentation or is it github rendering ?

File: frame/tips/src/tests.rs [hunk 10]
```diff
@@ -160,8 +160,11 @@ pub fn new_test_ext() -> sp_io::TestExternalities {
 		// Total issuance will be 200 with treasury account initialized at ED.
 		balances: vec![(0, 100), (1, 98), (2, 1)],
 	}.assimilate_storage(&mut t).unwrap();
-	pallet_treasury::GenesisConfig::default().assimilate_storage::<Test, _>(&mut t).unwrap();
-	t.into()
+
+	pallet_treasury::GenesisConfig::<Test, _>::default()
+		.assimilate_storage(&mut t).unwrap();
+
+    t.into()
```
Comment (thiolliere) [inline frame/tips/src/tests.rs:167]: should be tabs no space

File: frame/tips/src/tests.rs [hunk 11]
```diff
@@ -476,8 +479,11 @@ fn genesis_funding_works() {
 		// Total issuance will be 200 with treasury account initialized with 100.
 		balances: vec![(0, 100), (Treasury::account_id(), initial_funding)],
 	}.assimilate_storage(&mut t).unwrap();
-	pallet_treasury::GenesisConfig::default().assimilate_storage::<Test, _>(&mut t).unwrap();
-	let mut t: sp_io::TestExternalities = t.into();
+
+	pallet_treasury::GenesisConfig::<Test, _>::default()
+		.assimilate_storage(&mut t).unwrap();
+
+    let mut t: sp_io::TestExternalities = t.into();
```
Comment (thiolliere) [inline frame/tips/src/tests.rs:486]: same shouldn't be space

File: frame/treasury/src/tests.rs [hunk 12]
```diff
@@ -119,14 +110,36 @@ impl Config for Test {
...
+	pub enum Test where
+		Block = Block,
+		NodeBlock = Block,
+		UncheckedExtrinsic = UncheckedExtrinsic,
+	{
+		System: frame_system::{Pallet, Call, Storage, Config, Event<T>},
+		Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
+		Treasury: treasury::{Pallet, Call, Storage, Config<T>, Event<T>},
+	}
+);
```
Comment (thiolliere) [inline frame/treasury/src/tests.rs:126]: why did you moved it ?

File: frame/treasury/src/lib.rs [hunk 13]
```diff
@@ -303,46 +261,136 @@ decl_module! {
...
+		Burnt(BalanceOf<T, I>),
+		/// Spending has finished; this is the amount that rolls over until next spend.
+		/// \[budget_remaining\]
+		Rollover(BalanceOf<T, I>),
+		/// Some funds have been deposited. \[deposit\]
+		Deposit(BalanceOf<T, I>),
+	}
+
+	/// Old name generated by `decl_event`.
+	// #[deprecated(note = "use `Event` instead")]
```
Comment (thiolliere) [inline frame/treasury/src/lib.rs:310]:
```suggestion
	#[deprecated(note = "use `Event` instead")]
```

**Review by thiolliere — APPROVED** [review]
looks good to me, some indentation are wrong though

File: frame/treasury/src/tests.rs (hunk 12 above)
Comment (shamb0) [reply frame/treasury/src/tests.rs]: In some refernce test file saw that `construct_runtime!()` called after 
defining all trait configs. Revert back the position.

File: frame/tips/src/tests.rs (hunk 11 above)
Comment (shamb0) [reply frame/tips/src/tests.rs]: Corrected the bad intend

File: frame/tips/src/tests.rs (hunk 10 above)
Comment (shamb0) [reply frame/tips/src/tests.rs]: Corrected the bad intend

File: frame/tips/src/tests.rs (hunk 9 above)
Comment (shamb0) [reply frame/tips/src/tests.rs]: Corrected the bad intend

**Comment by thiolliere** [comment]
some deprecated usage of RawEvent must be fixed: https://gitlab.parity.io/parity/substrate/-/jobs/870164

**Comment by shamb0** [comment]
Fixed "some deprecated usage of RawEvent"

**Comment by shawntabrizi** [comment]
@thiolliere @shamb0 you must be careful here.

Parts of treasury, the Tips and Bounties sub-pallets use an old Storage Key, which would mean some kind of migration is needed. Maybe not relevant to this exact PR, but def something to watch out for.

**Comment by thiolliere** [comment]
yes tips and bounty pallets will need migration.
For treasury both old pallet and new pallet use the prefix "Treasury" in kusama and polkadot.

**Comment by shawntabrizi** [comment]
@shamb0 this also went stale

Feel free to reopen if you update to master

**Comment by thiolliere** [comment]
needs to merge master, and awaiting for another review.

I should also review the merge of master once done.

**Pull request closed without merging**